src/main/java/uk/gov/hmcts/ccd/domain/service/DocLinksRestoreService.java

```diff
@@ -0,0 +1,134 @@
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.ccd.data.casedetails.CaseAuditEventEntity;
+import uk.gov.hmcts.ccd.data.casedetails.CaseDetailsEntity;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.*;
+
+@Component
+public class DocLinksRestoreService {
+    private static final Logger LOG = LoggerFactory.getLogger(DocLinksRestoreService.class);
+
+    public static final String START_TIME = "'2019-08-20 15:50:00.000'";
+    public static final String END_TIME = "'2019-08-21 13:30:00.000'";
+    public static final String BETWEEN_CLAUSE = "BETWEEN  " + START_TIME + " AND " + END_TIME;
+
+    public static final String OLD_CASE_QUERY = "SELECT * FROM case_data WHERE created_date < " + START_TIME + " AND id IN " +
+        "(SELECT distinct(case_data_id) FROM case_event WHERE created_date " + BETWEEN_CLAUSE + ")";
+    public static final String OLD_CASE_QUERY_WITH_JUIDS = OLD_CASE_QUERY + " AND jurisdiction IN :jids";
+
+    public static final String LAST_EVENT_WITH_VALID_DATA = "SELECT * FROM case_event WHERE id IN " +
+        "(SELECT max(id) FROM case_event WHERE case_data_id IN :caseIds AND created_date < " + START_TIME + " GROUP BY case_data_id)";
+
+    public static final String NEW_CASE_QUERY = "SELECT * FROM case_data WHERE created_date " + BETWEEN_CLAUSE;
+    public static final String NEW_CASE_QUERY_WITH_JUIDS = NEW_CASE_QUERY +  " AND jurisdiction IN :jids";
+
+    public static final String NEW_CASE_EVENTS = "SELECT * FROM case_event WHERE case_data_id IN :caseIds";
+
+    public static final String DOCUMENT_URL = "document_url";
+
+    @PersistenceContext
+    private EntityManager em;
+
+    public List<CaseDetailsEntity> findDocLinksMissedCases(List<String> jurisdictionList) {
+        // missing links in old cases due to an edit during bug window
+        List<CaseDetailsEntity> oldCases = findDocLinksMissedOldCases(jurisdictionList);
+        // missing links in newly created cases during bug window
+        List<CaseDetailsEntity> newCases = findDocLinksMissedNewCases(jurisdictionList);
+        logStats(oldCases, newCases);
+        return ListUtils.union(oldCases, newCases);
+    }
+
+    private List<CaseDetailsEntity> findDocLinksMissedOldCases(List<String> jurisdictionList) {
+        Query query = jurisdictionList.isEmpty() ? em.createNativeQuery(OLD_CASE_QUERY, CaseDetailsEntity.class) :
+            em.createNativeQuery(OLD_CASE_QUERY_WITH_JUIDS, CaseDetailsEntity.class).setParameter("jids", jurisdictionList);
+        return findInOldCasesBasedOnEventHistory(query.getResultList());
+    }
+
+    private List<CaseDetailsEntity> findDocLinksMissedNewCases(List<String> jurisdictionList) {
```

Comment:
minor. feel free to ignore. we could also reduce even more the scope by passing case type rather than jurisdiction.